src/main/kotlin/no/nav/familie/ks/sak/kjerne/behandling/HenleggBehandlingService.kt

```diff
@@ -0,0 +1,98 @@
+import no.nav.familie.ks.sak.api.dto.ManueltBrevDto
+import no.nav.familie.ks.sak.common.exception.Feil
+import no.nav.familie.ks.sak.common.exception.FunksjonellFeil
+import no.nav.familie.ks.sak.config.FeatureToggleConfig
+import no.nav.familie.ks.sak.config.FeatureToggleService
+import no.nav.familie.ks.sak.integrasjon.oppgave.OppgaveService
+import no.nav.familie.ks.sak.kjerne.behandling.domene.Behandling
+import no.nav.familie.ks.sak.kjerne.behandling.domene.BehandlingRepository
+import no.nav.familie.ks.sak.kjerne.behandling.domene.BehandlingStatus
+import no.nav.familie.ks.sak.kjerne.behandling.steg.StegService
+import no.nav.familie.ks.sak.kjerne.brev.BrevService
+import no.nav.familie.ks.sak.kjerne.brev.domene.maler.Brevmal
+import no.nav.familie.ks.sak.kjerne.logg.LoggService
+import no.nav.familie.ks.sak.sikkerhet.SikkerhetContext
+import org.springframework.stereotype.Service
+import org.springframework.transaction.annotation.Transactional
+
+@Service
+class HenleggBehandlingService(
+    private val stegService: StegService,
+    private val featureToggleService: FeatureToggleService,
+    private val brevService: BrevService,
+    private val oppgaveService: OppgaveService,
+    private val loggService: LoggService,
+    private val behandlingRepository: BehandlingRepository
+) {
+
+    @Transactional
+    fun henleggBehandling(behandlingId: Long, henleggBehandlingDto: HenleggBehandlingDto) {
+        val behandling = behandlingRepository.hentBehandling(behandlingId)
+        validerOmBehandlingKanHenlegges(behandling, henleggBehandlingDto)
+
+        val henleggÅrsak = henleggBehandlingDto.årsak
+        // send henleggelse brev
+        if (henleggÅrsak == HenleggÅrsak.SØKNAD_TRUKKET) {
+            brevService.genererOgSendBrev(
+                behandlingId = behandling.id,
+                manueltBrevDto = ManueltBrevDto(
+                    brevmal = Brevmal.HENLEGGE_TRUKKET_SØKNAD,
+                    mottakerIdent = behandling.fagsak.aktør.aktivFødselsnummer()
+                )
+            )
+        }
+
+        // ferdigstille oppgaver
+        oppgaveService.hentOppgaverSomIkkeErFerdigstilt(behandling).forEach {
+            oppgaveService.ferdigstillOppgaver(behandling, it.type)
+        }
+
+        // lag historikkinnslag
+        loggService.opprettHenleggBehandling(
+            behandling = behandling,
+            årsak = henleggÅrsak.beskrivelse,
+            begrunnelse = henleggBehandlingDto.begrunnelse
+        )
+
+        // henlegg behandling steg
+        stegService.henleggBehandlingSteg(behandling)
+
+        // muterer behandling resultat og status
+        behandling.resultat = henleggÅrsak.tilBehandlingsresultat()
+        behandling.status = BehandlingStatus.AVSLUTTET
+
+        // oppdater behandling
+        behandlingRepository.saveAndFlush(behandling)
+    }
+
+    private fun validerOmBehandlingKanHenlegges(behandling: Behandling, henleggBehandlingDto: HenleggBehandlingDto) {
```

Comment:
Trenger du å sende inn hele dto'n her?
Kan man ikke bare sende inn årsaken?

```suggestion
    private fun validerOmBehandlingKanHenlegges(behandling: Behandling, henleggÅrsak: HenleggÅrsak) {
```